### ISSUE TITLE
Shelf & Multiple Hpub support

### ISSUE THREAD

**Issue by segundoseis** [Simbul/baker#20]
Was: more than one hpub in the same application

As a user would be great to see more than one hpub in the same application.
Pre-home with hpub-thumbsnails navigation?

**Comment by folletto** [comment]
That's a great thing, but I'm wondering if it's within the objectives of this tool, or we should just build the "framework" to be included in any other kind of app.

Still, I'd like to give some sort of basic functionality, so I'd like to hear some feedback about this topic. :)

**Comment by segundoseis** [comment]
I do not know if I'm complicating this. but I imagine an author who has multiple numbers, it would be interesting not generate more than one app in the appstore.
What do you think?

**Comment by folletto** [comment]
I think that two different situations exist:
- allowing to read "any" book (iBook)
- allowing to create multiple numbers

Because those two I think that are similar technically, but very different from the point of view of UI and author. :)

**Comment by segundoseis** [comment]
truth!
mmm .. now I'm thinking that I could download a zipped hpub as @svdgraaf was preparing. :-P

**Comment by MacBoy** [comment]
For those who would like to use Baker as a framework for magazines, each hpub would be a publication of the magazine.  The same for a product catalog, each hpub would be a different product or section of the overall catalog.

It would be very nice if the Baker Framework could hold multiple hpub packages. 

Thank you!!!

**Comment by aldriq** [comment]
I guess the main question is whether BakerFramework can/will evolve (or branch out) to a full-blown BakerReader that can hold any number of books, like iBook, kindle or Stanza.

**Comment by folletto** [comment]
Good point aldriq, that's truly something that we are thinking of, the point for that is the effort to reach that point, and what does it mean for the project, because then it would be a Reader with inside the Framework. Completely doable, but still decisions to take and development (a lot) to do. We'll see how the community is going to respond, step by step I think. ;)

**Comment by MacBoy** [comment]
Now that we can replace the internal "book" with an externally downloaded book, the contents can easily grow and be replaced.  The HTML could have an Index page that one can link to different sections.  That would be the workaround for having only one book in the app.

We still have a need for multiple books within the app.

Priority for it is lower than other features.  For example, the Landscape orientation should be done first.  This request has a suitable workaround already.

**Comment by aldriq** [comment]
I understand, it's definitely a medium/log-term decision. I imagine it also depends on how 'fast' the next hpub standard takes to consolidate.

**Comment by therabidbanana** [comment]
I'm planning on adding multiple hpub support to my Baker fork for a project I'm working on. Is there any planned movement on this from the Baker team, or planned updates to the hpub specification that might be of interest (the recent addition of a JSON library by @folletto seems to hint at moving toward JSON manifests)? If there are plans to support metadata in manifests that could be quite useful.

Here's what I'm thinking of building: 

Keep every hpub downloaded, and the associated book:// url, with a dictionary mapping url to hpub. On a book:// url, check to see if it's already been downloaded by looking it up in the dictionary. If so, just unzip the local hpub, else download as normal. 

For some people, that might be enough, but for flexibility I imagine I will also need to add the following capabilities: 
- Showing the user a list of hpubs already downloaded.
- Allow deleting existing local hpubs
- Allow updating local hpubs (transparently delete and then redownload if versions don't match).

I see two possible ways to handle the extra use cases:
1. (Use Javascript/HTML as much as possible) Inject the dictionary of local hpubs into the page as a javascript variable. Either inject into every single page, or maybe only book://local pages, or maybe only a specific file defined in the application (index.html might be a good place). Add support for book://local/path/delete so that books can be deleted. Let users define whatever javascript they'd like to take the local hpub list and work with it. The injection of javascript variables seems kind of hackish, but would allow a lot of customization with just standard html/js.
2. (Obj-C all the way) Create a new UIView accessible either by a new url like book://shelf, or maybe in place of book://local if preferred - maybe build on top of something like https://github.com/AlanQuatermain/AQGridView to give a pretty interface. This option seems less hackish, but is also a lot less flexible.

Anyone have any thoughts on which version they'd prefer, or things I might not have considered?

**Comment by folletto** [comment]
Wonderful, let's see. :)

You're absolutely correct about the features needed in order to allow proper "download-and-storage" feature, since if we store something we also need to give a way to manage it (select, delete, update).

For that, I was thinking about a solution that involved the creation of a top navigation bar when you double tap. The bar is going to have a button "All books" and will allow you to tap on it and get a dropdown with all the downloaded book in the issue. From there, you can select one. If it was a downloaded book, on there it will also provide an update feature and a delete feature, even if it's not yet clear to me what's the best way to build it. As before, we probably have to create it, see how it works, and iterate a couple of times to get it right.

But... this doesn't mean that a HTML integration is excluded. We should still give ways from the book itself to work with book urls, but I don't expect the ability to delete/manage them, just to choose them. And for that a JS injected would be fine, let's just coordinate on the correct namespace for it. Probably creating a object called "book" is ideal, and we can append everything to it, so: book.shelf that returns an array of hashes: [{title, url}, {...}, ...].

You also got right the direction we are taking for the manifest. Again, it's more in the experimentation phase right now, so it might be that we scrap it, but we are testing something what uses a "book.json" file inside the hpub book, containing things like title, author, reference URL, pages and sheet options (orientation, size). I hope to be able to share something soon inside the example book that I'm re-building. :)

**Comment by therabidbanana** [comment]
I had considered adding a navigation bar, but thought it could get kind of cluttered showing even more on double tap, and I wasn't sure how minimalist you were hoping to keep the interface. As long as it's easy to disable, I suppose that's not an issue though.

I like the way the dropdown works in Dropbox (http://c0736882.cdn.cloudfiles.rackspacecloud.com/Dropbox%20iPad%20-%203.png). Something similar to that may be appropriate, but I have no idea how hard it is to program.

One thing I hadn't really considered - it sounds like you're imagining being able to see books in the list that haven't been downloaded yet. How do you imagine letting the application know those books exist - hardcoding the list into the app? Maybe some sort of JSON list or RSS feed pulled off of a server? My original thought was just having the local book be a simple, one-page hardcoded list of book:// urls and releasing an app update every time the list of books (or in our case, magazine issues) changed. That's probably not very scalable though. It might be better if there was a list the application could retrieve from the server - but then we'd need to work out a standardized format to use for listing books - maybe just a JSON array of manifests?

**Comment by folletto** [comment]
Yes, I'm imagining something similar to that UI. :)

About providing view of books not downloaded yet I wasn't really thinking anything, we might leave it for a future release. I won't put too much work ahead, or we wouldn't get a good feedback from all the users. ;)
The one page approach you propose however should work well, and straight out of the box: if a book references another book, it should be easy to download it and add it to the Baker reader, so you could have a "starting" page, and then any other book will be able to trigger the download of other books. :)

Btw, with these features it's probably possible to release a Baker Reader, so... that's something we were thinking of, but with your help it might get here sooner than expected. ;)

**Comment by therabidbanana** [comment]
I have a beginning implementation working in https://github.com/orangesparkleball/baker/tree/multiple-hpubs.

A navigation bar shows up on double tap with a popover for displaying the list of stored books. The implementation is still a little crude - each hpub must have a unique filename, and updates are not download if the filename is the same. There's also not any way to delete books yet.

**Comment by folletto** [comment]
Excellent! I don't have time right now to try it out, but I'll do asap with the last master from that tree. :)

**Comment by folletto** [comment]
Tested :)

Ok, my proposal would be (sorry if I'm going to say even polish things and things you might already have planned):
- Move the button on the right, like Mail and Dropbox
- Use the book name instead of baker in the top bar
- Use the book name instead of the filename in the list (this probably needs the book.json manifest)
- Polish: fix a bit the animation in order to be coherent with the status bar

I omitted things like delete and update, since we already talked about it. ;)

**Comment by therabidbanana** [comment]
Was just about to post an update on this front. :)
- Do you mean putting this button on the left instead of the right? I was thinking we might also add a "Close book" arrow and place it on the left.
- This is now pulled from the manifest (if one exists).
- A bit trickier - since currently manifests are zipped inside the hpub, brings up a question below.
- I was having trouble trying to get them to come in together in a way that didn't look haphazard. I'll play with it a bit more.

Question - does it make more sense to keep the hpubs, or the extracted versions of the hpubs? Theoretically the hpubs will take less space, but I can't imagine it's a significant enough amount to worry about. On the flipside, keeping the hpubs makes them easier to treat as discrete books so you can move books into Baker with the iTunes File Sharing system. 

If we keep them around as hpubs, it's probably worth caching any manifests into a database so the system doesn't need to extract the hpub to find the manifest - although I think the zip library might have support for peeking into the zip file without extracting. Maybe that's an option. Will need to investigate.

**Comment by folletto** [comment]
1. Yep. Uh, what were you thinking then happens when you close a book? I would take a simpler approach: in the dropdown there's also the "master" book (the first one) so you can simply swtich back to it. (this will open the system to more flexibility in the future as well I think).
2. Great!
3. Yes, good point. Regarding that question, I think you detailed well the possibilities. I'd try to keep the zipped book as well, exactly for the same reasons you just explained. Then, I'd just store somewhere the manifest (or in a db) in order to use it. The only drawback is the load time for opening a book, but it's something that we could check.
4. Maybe we can just wait the opening of the bar when the status bar is completely folded out, and then start animating in our bar. But it's just a thought, if you find a better and faster solution go for it. ;)

**Comment by therabidbanana** [comment]
The dropdown can now peek into the hpub and look for a manifest. If it finds one, it replaces the filename with the title from that manifest, otherwise it just uses filename. With this method we don't necessarily have to store manifests separately. 

I think we still might want to store manifests separately - that'd allow you to delete a book but keep a reference to it so you could easily redownload it... but I imagine peeking into an hpub to find the manifest could still be useful if you downloaded a book but didn't open it (going back to the iTunes file sharing method of adding books), so it's a good first step.

**Comment by folletto** [comment]
I agree with the approach, great. 

For the filename we might think of using a hash of the URI (since that is going to be the unique identifier, like IBAN) or use a physical path that matches the URI itself (for example: book://example.org/publisher/author/book -> /books/example.org/publisher/author/book.hpub).

**Comment by therabidbanana** [comment]
Is it worth forking this effort off into its own "baker reader" project? Our requirements for this project have evolved to create a "like iBooks" reader, complete with a shelf and the ability to download multiple books. I've got some of this going in my multiple-hpubs branch, but I'm restructuring so much of the code it's starting to look like a completely different program and it's not making sense to even support a single bundled book any more. 

My end goal is to release to the App Store a "Baker Reader" program that is basically an iBooks clone with hpub support instead of epub. Then, rather than having to build a new app for each of our clients, we can have them download that and then send them the hpub via email or book:// url. I get the reasoning behind creating per-book apps - so each one can have its own unique look and feel, and be a separate app in the  app store. So I can see wanting to have both a "Baker Core" that works as Baker does now, and a "Baker Reader" that pulls in elements of the Baker core while adding more of the iBooks like functionality.

**Comment by folletto** [comment]
Well it depends, I think that it should be possible to keep the two things together for quite a bit, so, what feature do you think it's going to break it?

I was trying to test your latest branch but it's giving me a few issues (NavBarController.h not found). I'd like to test it and to see what you want to add so we can coordinate.

That's because the Book Reader in many ways is going to be exactly Baker for a single magazine with multiple issues, with the OS-level book:// support disabled, at least thats's how we were planning it. :)

**Comment by therabidbanana** [comment]
Whoops, I did a lot of restructuring and didn't push all of my changes. Just pushed another commit or two that should remove any last references to the NavBarController.h.

As of yet, nothing I've done specifically breaks Baker, though some of the changes might be hard to merge (I split out some new classes to help cut down on the code in the RootViewController). 

I just wonder if the changes I plan on building (no bundled book, a shelf view for viewing hpubs, the navigation bar and book:// url support) are getting to the point where they warrant splitting the project between a "Baker Core" and a "Baker Reader", where the reader duplicates the Core classes to handle viewing books, but also adds in a Shelf view and hpub handling support, etc - stuff that wouldn't make sense for people building a simple, single-book app. 

I don't know if it'd make sense to keep adding constants to turn on and off extra features, which is probably what we'd need to do if they were both contained within the same repo (and include two different plists for building the app depending on if you want Reader or Standard - since url schemes have to be added with the plist, etc). I suppose they could be separate branches, but if they are maintained as two separate branches then that's not really any different from forking the project.

**Comment by folletto** [comment]
Ok, but I don't see at this time a breaking feature. :)

I mean: the only difference seems that you don't want a bundled book. It's quite simple to add a flag "if a book is present in the shelf, start with that". That will work for either single books, multiple magazine issues, series and Reader as well. :)

If you just want one book, you are just going to have a button that takes you to a shelf with only one book. No big deal (and even that could be a flag).

I agree that at some point the fork might happen, but I don't think it's now. Or am I missing something? :)
(for the book:// protocol, it should be just off by default on OS level, to avoid overlapping features, there should be only one reader). :)

Also: the navigation bar is already in the master branch, are you thinking something different? :)

**Comment by therabidbanana** [comment]
In my mind, one of the key features is a reader that _does_ have book:// turned on at the OS level so that I don't have to distribute apps via testflight every time I want people to be able to preview a new book (so far we're building these books as internal enterprise apps, not general public apps). Being able to email book:// links to distribute new copies is ideal (or include email attachments). We want to build several books and know our clients can view them without going through  the hassle of getting us  their device ids and without having to deploy to the app store, since the books would be private. 

I suppose if people are fine with a shelf view for their single-book apps that's fine, though the shelf viewing support might be adding a decent amount of code to the project (unsure if it'd significantly impact performance to have the code there and just ignore it - but it seems silly to force users to have the shelf if they don't plan on using it). 

I should note that I plan on adding support for multiple "shelves" of content, where each magazine/collection would have its own shelf within the reader and be able to notify the reader when there is a new book within the shelf (maybe each shelf is an RSS feed of links to hpubs or something like that?), on top of the ability to manually add books to a shelf. So in my mind the shelf isn't just a tiny bit of extra code, it's going to add a lot of extra functionality that just doesn't make sense for single-book apps.

Our end goal is to provide our clients with a CMS (based on the bakers-oven idea I've been toying with) for creating new hpubs (or updating existing ones) and distributing those hpubs to a set of users in the quickest and easiest way possible.

**Comment by folletto** [comment]
Could we get there in steps? Because maybe multiple shelves isn't going to be part of Baker Core, but everything else will. It's a hugely requested feature, it _is_ going to be part of Baker Core. :)

After that, I think that a proper forking will be a good choice. :)

**Comment by therabidbanana** [comment]
Okay, I wasn't sure if the plan was to include support for the shelf in the Baker Core. I definitely plan on getting a single shelf working before trying to work on the multiple shelves idea, hopefully it'll be ready within the next week or two.

**Comment by folletto** [comment]
Great. :)

Do you already have something in mind? Or are you starting with a plain shelf? I'm asking because I've already started working on the design of that part, and I'd like to avoid duplicated efforts. ;)

**Comment by ravitola** [comment]
Hi there

Checking on this discussion to see if there has been any progress to move this along as I would be interested in collaborating on this effort.  Let me know..

**Comment by therabidbanana** [comment]
Hey @ravitola, no real progress since my last comment in June, unfortunately. I've been a bit sidetracked lately. You can check out the work so far in my topic branch: https://github.com/orangesparkleball/baker/tree/multiple-hpubs

Still to be done:
- Fix the index view so that it isn't pushed down by the navigation bar
- Create a better interface for the list of hpubs

What's been done so far:
- A large amount of refactoring to create "book" and "shelf" models and separate view controllers responsible for different parts fo the app
- Enabling of book:// url catching from external apps
- Enabling file-type recognition so you can "Open With..." the file from other applications such as Mail and Dropbox
- Enabling file synchronization in iTunes so you can add hpubs with iTunes

**Comment by folletto** [comment]
I'd just suggest in any case to disable the book:// url catching until we have a bigger discussion about it, I'd avoid to release frameworks then are going to compete on which one is going to download books - and maybe you don't want to download any kind of book in your branded app, you know ;)

But that's critical instead to create a reader, so worth working with that, it just needs to be off by default. :)

**Comment by therabidbanana** [comment]
Well, I wanted to build support for catching the URL scheme, even if it does get turned off by default, so in my testing branch it's on. To turn it off is an easy plist change though.

**Comment by folletto** [comment]
Yep, we did the same, I just wanted to avoid situation of apps published on the store from the evolved framework that then are going to catch all the books from everywhere. :)

**Comment by therabidbanana** [comment]
Hey everyone, just thought I'd give a quick update since I got a message in my Github inbox today:

I've stopped working on this and I'm not sure when I'll have the chance to get back to it any time soon. I recently switched jobs, so haven't been able to make changes during my work hours any more. If anyone wants to take the lead on this, feel free to take a look at what I've done so far in my branch and either build off of it or start fresh (I did a pretty intensive code refactoring, so it'll probably be nearly impossible to merge into the latest code). 

I'll probably loop back to this at some point when I have a bit more free time, but I definitely won't have any time in the next few months to work on it, and if it's done by then I'll just find something else in Baker to fix. :D

**Comment by folletto** [comment]
Hi! :D

Even if it isn't a "good" news, thanks for the update! :)

We are going to tackle it in the 4.0 release, so we are going to have a look at your work by then, and see if we can borrow idea. 4.0 will be probably also a refactored version, so... we'll rewrite a lot of it as well. :)

Thank you very much, and I hope you'll still find a few minutes every now and then to chime in. :)

**Comment by mpena2099** [comment]
Hi folks!

Please, see this: http://ios-blog.co.uk/iphone-development-tutorials/how-to-make-a-magazine-app-in-ios-part-ii/

Thanks!

**Comment by folletto** [comment]
Thanks! That's a very nice link. Did you try to contact him to see if he wants to try an integration with Baker? ;)

**Comment by mpena2099** [comment]
Not yet, but I'll try! :)

**Comment by viggiosoft** [comment]
Hi,
I'm the person that prepared the articles about magazine in http://ios-blog.co.uk and http://www.viggiosoft.com . I have been contacted to follow this project (I didn't know it before, excellent work by the way, congratulations!) and provide my cooperation in this discussion. 
It seems to me that everybody agrees on the need to extend the framework by adding the front-end (shelf) and back-end (client/server communication, Newsstand, In App Purcahse). While there is no a clear direction if these implementations should be added in the current framework as it or as new frameworks. 
I have prepared several apps, and on of my great early choices was to implement a reusable PDF reader (with all possible options: search, links, multimedia, ...) in order to be able on each app to re-use 80% of the reader (customer will require a customization) and the need to build just the shelf (front-end). Later I found that having 3 parts is better than two: the reader, the front-end shelf (UI), the back-end. In most cases the customer will ask you to rebuild from scratch the UI, while the back-end is highly reusable in different apps and different devices. 
So my opinion in such case is to avoid extending the existing framework, excellent as a reader, by adding back-end features and UI views. My suggestion is to provide other two modules that will link with the reader and will provide some guidelines for custom implementations. E.g. the back-end will change its implementation based on the kind of server it needs to work with or on the protections applied to provide In App Purchase, but the download and installation of the content will be done in the same way as this is established by the reader format. Besides the front-end will contain the basic hooks or protocols to both the reader ("open this book") or to the back-end ("purchase this book", "download this book") but it will be up to the final developer customize this part of the app.
Anyway I would be happy to provide my contribution based on the directions that the owner of this project will give to me.

Carlo

**Comment by folletto** [comment]
Hi Carlo! :)

At the beginning more than a year ago (_wow_, thinking about it) we didn't have in mind to expand the framework to such a level, so the code isn't much modular.
It would be great to have a structure like the one you defined, but it's probably related to the refactoring tracked in Issue #59 that still hasn't been done.

If you want to contribute to Baker, the best idea would be probably for you to have a look at the code base and suggest here an approach that could be doable in your perspective. We're open to discuss any option. :)

A cool approach could be for you to work on a branch from the upcoming 3.1 release and incorporate in your shelf, discussing and feeding then back the result. But, I think, this could be quite a big undertaking, so I can't really ask you to do it. :)

Check the code and let's see what you think you could do. :)

Thanks! :)

**Comment by viggiosoft** [comment]
Hi Davide,

thanks for your suggestion. Definitely forking 3.1 and trying to refactor it in such way that it can talk with an external shelf using a general purpose interface is a great idea. I will try to experiment a little bit privately in order to have a deep understanding of the code sections I need to interact with and if I get some valid result I will  be happy to publish and explain my comments.

Thank you

**Comment by folletto** [comment]
Thanks to you! Please keep us informed on your thoughts and progress. :)

**Comment by mpena2099** [comment]
Hi!

First of all, sorry Google by his translator. :D

I contacted Carlo because I have to build an iPad digital magazine similar to existing ones here in Brazil and, as I would use the Baker, the idea was to join the code of Carlo with Baker (I still do not know if the Carlo's Shelf is BSD License but... =D)

I am new to developing to iOS/Xcode but I'm a WEB developer for years.

Despite the difficulty, I will try to put Baker on Carlo's Shelfr. And place HERE the results, of course. ;)

OK?

Thanks!
Mauricio

**Comment by folletto** [comment]
No worries, and thank you in advance for any contribution you'd like to make. :D

**Comment by mpena2099** [comment]
Hi!

That's what I'm trying to do to pub Baker on "Carlo's Shelf":
1. change the file links to download store.plist HPub COMPRESSED files;
   (DONE!)
2. download OK, unzip the file using the class decompressing already exists in Baker;
   (DONE!)
3. Save journal in a directory created with the name "ID" magazine;
   (Carlo's Shelf already do something like this, so... DONE!)
4. hPub file unziped, display it in the same window the Carlo's Baker display the PDF file, ADDING Baker code;

OK... THAT is the problem! No idea how to do this! =D

Above is the method (on ShelfViewControleler.m) of Carlo's Shelf that reads the file:

---

-(void)readIssue:(Issue *)issue {
    QLPreviewController *preview = [[QLPreviewController alloc] initWithNibName:nil bundle:nil];
    preview.delegate=self;
    preview.dataSource=self;
    urlOfReadingIssue=[[issue contentURL] URLByAppendingPathComponent:@"magazine.pdf"];    

```
[self presentModalViewController:preview animated:YES];
```

## }

QUESTIONS:
1. it's possible to call Baker INSIDE this method???
2. how to do this? :P

Thanks for all!

**Comment by viggiosoft** [comment]
Hi Mauricio,

can you share your work as a fork on GH? I will try to have a look and see how can I help you on point [4]. Honestly I had no time to look at the whole stuff, all my customers are pushing me to see all they apps released by Christmas and I'm really busy... But as you already merged the shelf with baker in a single project, this can simplify my job.

Carlo

**Comment by mpena2099** [comment]
Hi Carlo!

Yes, I'll share my work, no problem. I don't know how to use GIT yet but I'll do this today and post it on GH. Any help after that will be welcome. ;)

Thanks!
Mauricio

**Comment by mpena2099** [comment]
Hi folks!

If I made this correctly, see the Baker + HowToMakeAMagazineApp together in https://github.com/mpena2099/bakershelf 

Now I need a help to call Baker from HowToMakeAMagazineApp...

Thanks!
Mauricio

**Comment by viggiosoft** [comment]
Hi Mauricio, I just opened an issue about some missing files in your fork'd project. I would keep all issues and communications related to that experiment on the new bakershelf git so people interested can follow it in the right place.

**Comment by nin9creative** [comment]
Has anyone seen this example?  http://www.siteless.org/?p=585

**Comment by mpena2099** [comment]
Thanks Carlo, I'll fix it tomorrow, and post related communications in the "bakershelf" GIT.

nin9creative... I have to test it. =D

**Comment by viggiosoft** [comment]
@nin9creative It seems this example is what we're looking for. I checked-out its pugpig integration and it works very well. The code is quite clear and it seems they're working in integrating Baker.

**Comment by folletto** [comment]
@nin9creative the last time we checked, it wasn't integrating with Baker, but it might have changed. Worth testing. :)

**Comment by nin9creative** [comment]
I emailed the guy that wrote the code to see if he wanted to join the conversation.  I actually haven't integrated the two yet but I poked around the code - pretty well done.  I also tweaked some of the look to mimic the Adobe Digital Publishing look and feel just for a test.  See a partially obscured version here.  http://public.nin9creative.com/baker/BakerShelf.png

**Comment by mpena2099** [comment]
The big question is: how do I create/call a baker instance from some method of other class????

**Comment by nin9creative** [comment]
I think the root issue here is how do we structure baker so that it can be invoked as the reader for a given hpub.  That way people can create different sorts of shelves or whatever to launch a specific downloaded book / issue.  I think that many of us are looking for some sort of Magazine / Issue shelf like many of the publishers using the Adobe Digital Publishing Suite of tools (which is insanely expensive for indie publishers).  

The Siteless magazine shelf above is the closest that I've seen and in my mind worth developing / working on further.  The issue is coming up with an easy way to launch Baker to view any given hpub...  @folletto has there been any work in taking the current Baker implementation and making it easy to integrate with shelf like apps?

**Comment by folletto** [comment]
@nin9creative Thanks for the question, I'll take it as an occasion to clarify a few points. :)

# Where are we now

First of all: this very issue where you are discussing is the very place where we are collecting feedback and where we'll post updated about the development. So far, we did a lot of discovery, but since we are still fixing bugs and adding features for the 3.1, no code has been written yet.

Baker _will_ support a shelf, and as you can see at the top of this page it's scheduled to be Baker 4.0. We are still not sure on the specific implementation however, and that's why we are both looking into different solutions and following this topic as well to see the different opinions of the people working on it.
We are looking for a solution both flexible and without administrative debris as much as possible.
(see Tufte for a definition of administrative debris: http://www.edwardtufte.com/bboard/q-and-a-fetch-msg?msg_id=00036T).

Now, there are of course two different things here that are possible:
1. We abstract Baker as a **class** — currently, for historic reasons, it's an app not easily pluggable (it's tracked by Issue #59)
2. We **include** in Baker a shelf as flexible as possible (it's tracked by this issue).

Both the solutions are valid, and in the end _we'd love to do both_. Right now however, we see more feasible doing the second one, because the first one will require a complete code refactoring. However, no decision is final, as of today, and we'll keep following this thread. :)

Plus, it's an opensource project, so someone might fork Baker and implement one of the two points above and feed it back to us... ;)

**Comment by nin9creative** [comment]
@folletto thanks for the clarifications.  I'm excited about this.  As soon as Baker can be used as a viable alternative to some of the major software solutions for doing newsstand subscription type stuff, I think you'll see a huge uptick in people interested in baker and what it can do!  Very exciting.

**Comment by mpena2099** [comment]
I think I need to study more OOP and Objective-C to try help Baker Shelf...

No idea where to start to make Baker a class. I guess it was! =D

**Comment by nin9creative** [comment]
**Consolidating multiple posts that I had on another issue thread to this more appropriate thread**

So I totally hacked this together to see if I could get it working. It's kind of messed up because I lost the Interceptor that catches the double taps and that sort of thing. Didn't look much into it - however I took the Siteless.org sample and combined the base code that they were using for the shelf with Baker. I got it working where I could download the issues from my server and open them in baker via the shelf view. It is totally hacky, but I was just trying to get the dumb thing to work! I think I almost pulled my hair out figuring this out.

Here's a video of how it is working right now. I just am slapping a button on the webview of baker so that I can get back to the shelf - ugly yes :)

http://public.nin9creative.com/baker/BakerShelfDemo/

If that prior link doesn't work this is a direct link to the quicktime file from screenium.

http://public.nin9creative.com/baker/BakerShelfDemo/BakerShelfDemo-desktop.m4v

If you get the code from the "bakershelf" branch on my fork you can just run the sample. It will grab a list of issues from my webserver and also I've zipped up 5 versions of the default scarlet book with just a different title on the first book page so that you can see that it actually changes when you click on the covers. Not sure how this might play into a real shelf in Baker but I've been playing around and it looks promising. There is stuff missing such as a progress indicator on the actual content download, and a bunch of other stuff as well.

I made a generic version here - this is what it looks like. I also need to change the .json to include a price etc for display, right now it's just hardcoded.

http://public.nin9creative.com/baker/GenericBakerShelf.png

**Comment by nin9creative** [comment]
https://github.com/nin9creative/baker/tree/bakershelf

**Comment by mpena2099** [comment]
Great job Andrew!

Now what we have to do to finish it? :)

**Comment by mpena2099** [comment]
Andrew, let's put a list of what needs to be done in https://github.com/nin9creative/baker/. I will see which items I can finish.

**Comment by nin9creative** [comment]
Ok will get to it later...  There is a bunch that needs to be done :)

Warm Regards,
Andrew

**Comment by nin9creative** [comment]
Changed the home page of the wiki on my fork with a few items to look at.

**Comment by folletto** [comment]
As for the placement of the button, the ideal position that we tested as working well is to add a black bar on top when you double tap: in that way the full magazine/book experience will still be there, while at the same time you have access to the full shelf. :)

See this mockup I prepared for the shelf work:
http://cl.ly/2l0k3M0c04320x1e2e3l

In the design:
- The bar is full black
- The title of the book is Helvetica Light. :)
- The button is basically a flat, dark gray version of the standard button

I hope this helps. ;)

**Comment by folletto** [comment]
Just one question that just popped into my mind: what's the license of the library developed by siteless? Does someone know? I checked the website but I wasn't able to find anything specific.

The best thing would be to be BSD, aligned with Baker itself, to allow commercial reuse and implementations. :)

**Comment by mpena2099** [comment]
@nin9creative, how do I add "issues" on your fork, like here in Baker? That is the best way to create a "check"/todo list.

@folletto, they just said that "all code is gonna be open sourced" (http://www.siteless.org/?page_id=156), and I guess that BSD will be the best choice.

**Comment by folletto** [comment]
Then we have to ask to be sure. :)

**Comment by nin9creative** [comment]
I was trying to get ahold of the guy who posted it to make sure it was ok to use.  Haven't heard back.  Will try again!

Warm Regards,
Andrew

**Comment by nin9creative** [comment]
Cool

Warm Regards,
Andrew

**Comment by nin9creative** [comment]
Yes we should do that....

Warm Regards,
Andrew

**Comment by mpena2099** [comment]
I think the code from @viggiosoft (https://github.com/viggiosoft/HowToMakeAMagazineApp) will help putting a progress indicator on the "Siteless Shelf", and also use Newsstand...

**Comment by nin9creative** [comment]
I added the Issues option to the forked project...

**Comment by mutador** [comment]
mi comments about the look & fell of the version off nin9creative

my vision is like a fashion magazine ....

1- the first page , always a generic cover, whit alpha animation e.g
2-the look and fell mut be customizable for the user in html, embed a html not code in xcode ( in general let the user personalice his own page)
3- if the embed html is not posible, not use the look and fell of the sencha touch aplications, is too much tecnic and not fashion, the top and the footer whitout background ( background: none; ) and let the user to put an image like background. the  idea is make transparent buttoms and a .png fot the look&fell

**Comment by nin9creative** [comment]
Feel free to contribute to the code!

**Comment by mutador** [comment]
sorry, but i am designer whit very basic know of html, we are looking for someone in spain  for developer xcode framework like baker and not found nobody, we only help in desing and contributions 

but understand  of the idea of the background image for the page, is a god idea to customize the top and background

**Comment by nin9creative** [comment]
I think that this would be pretty hard to do as strictly HTML for one because of the nature of what is trying to be done - there needs to be hooks into the code to handle downloads, in app purchase, all sorts of stuff...  You are right though it should be slightly customizable by the end user - however to what extent I'm not sure.  I'll defer to @folletto

**Comment by mutador** [comment]
but i only say, dont use background in xcode for the top and the bottom and use a background image for the page ( 1024x768) , off course all the text of the title, price and xml in xcode.... you know where i mean

<img src="http://www.mutador.com/baker/personalize.jpg" width="768" >

of course this is not a priority...

**Comment by nin9creative** [comment]
That'd be great - pictures are always easier to understand! :)  Thanks.

**Comment by folletto** [comment]
The extent of customizability is something we can have after everything is in place and working, so let's reach that stage first. :)

As the Baker team, we were thinking of creating a Shell in full HTML with specific hooks and server-side downloads, but of course now that you are developing an alternative we'll check that out first. :)

The problem with native interfaces is exactly the extent of the customizability that's possible with a reasonable effort, so we can't look into that without first having it working fully and reliably.

So, please **open a new Issue** related to this one about that, we'll tacked it after this one. :)

**Comment by folletto** [comment]
One point that probably we should discuss instead here is: _what will open at the first execution?_
1. First execution opens the latest magazine
2. First execution opens the shelf

Both options have pros and cons, and at this point we are leaning toward opening the magazine first, because that's what the user expect, but I'd like to hear your opinion too. :)

**Comment by nin9creative** [comment]
I would look at how all the other Newsstand/Magazine apps work.  All of the big guys using the Adobe Digital Publishing stuff (Oprah / Martha Stewart / etc).  

They seem to always land on the shelf if you are opening the app from the Newsstand interface.  Some (not all) also seem to update the Newsstand icon with the latest available issue cover art.

However - opening the app always seems to land on the shelf view...

**Comment by mutador** [comment]
all the issues are integrated and upload in the newsstand, and there are a way ( but we don´nt found) to linkage this upload issues whit the visor...whitout a extrenal hosting .( bye bye adobe solution.. for hsoting)

this is for pdf
http://www.viggiosoft.com/blog/blog/2011/10/17/ios-newsstand-tutorial/

and this make the solution , but they host your magazines, and we think is not necesary if you know the link of the issue tha you upload by newsstand. 

http://urbanairship.com/

**Comment by folletto** [comment]
@nin9creative Thanks. :)

@mutador As far as I know, you can choose where to host your files. :)

**Comment by mpena2099** [comment]
@viggiosoft, if you could help us with the shelf integration... @nin9creative is doing a good job in https://github.com/nin9creative/baker

Thanks!

**Comment by kremisi** [comment]
Have you've ever seen the Inkling solution?
http://www.inkling.com/

You buy the content on the website and receive the issue on the app. In this way you'll skip to pay anything to Apple. This is a very interesting solution for publishers. Thanks, K.

**Comment by folletto** [comment]
Inkling is a _great_ application, both as a product and for the user experience and interface solutions it adopts. :)

**Comment by kremisi** [comment]
Have you ever thought to add an updating or buying function from external links?

**Comment by folletto** [comment]
What do you mean? :)

**Comment by kremisi** [comment]
It seems now you're focusing on storing, updating or selling multiple issues from within the shelf, I suggest to solve this feature outside the shelf, such as you can see in the Inkling example. And I think this is a great commercial improvement too in terms of AFS (Apple Fee Savings :-)

**Comment by folletto** [comment]
Well, the Shelf - this issue - is just a way to visualize multiple issue, nothing else. Or, in your words: storing.
For updating you can already use the `book://` protocol, and it's quite a flexible standard. :)
For purchasing, I don't think right now that Baker will support any specific solution. 

So, I think that we are perfectly aligned. :)

**Comment by kremisi** [comment]
Imagine my customer is visiting my website, ok. He decides to buy a new book, ok. He has bought the book and he receives a link such as book protocol book:// or http:// that will open the shelf on the iPad... Is this possible now?

**Comment by folletto** [comment]
Yes, of course. Just mind that:
1. The shelf is not ready yet, so any "update" will remove the previous book
2. You shouldn't use "book" as the protocol, that should be reserved for a generic reader application. Use a solution like Facebook ("fb87930283904://"), creating a long alphanumeric protocol identifier that's unique to your service and plug that in your iOS app.

**Comment by kremisi** [comment]
OK thanks a lot for your reply! Grazie, K.

**Comment by nin9creative** [comment]
@folletto Can you take a look at Issue 2 https://github.com/nin9creative/baker/issues/2 on my fork / bakershelf branch?  Just a quick description.  With the shelf code out of the box, the Baker (RootViewController) is no longer what is first loaded.  I actually moved the baker code that was there to a (ReaderViewController) class.  In any event along the way the Interceptor window functionality was lost.  I can't figure out how to add that back in on top of the push of the ReaderViewController class since the Interceptor was designed to sit over the main window view.  This is holding up some of the progress that I and @mpena2099 are trying to make because the tap functionality isn't working correctly :)  I for the life of me am hitting a brick wall with this one.

Not sure what you are working on or if you are doing something different for the shelf, but if you could help or if someone else that knows the Baker/Interceptor code well could take a quick look and see if it's an easy fix that would be great.

**Comment by folletto** [comment]
Ok, I'll comment there. :)

**Comment by josearellano** [comment]
I'm using BakerFramework in order to use multiple hpubs. I can download any hpub from aour server and update and manage them, but i have a problem, its about memory. For any reason I can not release the memory any time I close the reader, I debuged the application and I released it never use the Dealloc method. The application crash because of memory problems.  Thanks.

**Comment by folletto** [comment]
Did you try profile it? Because we constantly run the app with Instruments and so far wa aren't seeing any clear leak. Could you check what's causing the leak for you?

**Comment by josearellano** [comment]
Yes I tried it, and the problem seem to be that the application fill the memory. I don't see leaks. I'm using the RootViewController as another UIViewController, and I tried to add this viewcontroller with 2 differents ways, first adding this with [ presentmodalviewcontroller...] and the other one with a simple [ addsubview]. Both of them ways have a problem. When I try to close the view controller, for example in the case of a modal with [ dismissModalViewControllerAnimated...] the function Dealloc never its called, so the reader never cleans the memory, so after read some hpubs the application crash for memory space. Thanks a lot for your answer. :)

**Comment by folletto** [comment]
Uhm ok, but that's a restructuring you did. Baker - up until now - was never meant to be used in that way, so it's surely missing code to work as a standalone UIView class. You'll have to add it yourself.

**Comment by mpena2099** [comment]
Sorry to repeat it here but some people are getting making good progress in the creation of the shelf, but we need more help. Look at https://github.com/nin9creative/baker 

Thanks! =D

**Comment by nin9creative** [comment]
@pepereptio did you ever figure out your memory issues?  We're running into the same problem with our shelf...  Can't track down the memory leak.

**Comment by josearellano** [comment]
Yes, we had to change a little the code. in order to get to the dealloc method we had to present the view controller like a modal. Also we modified the way of getting the touches. And finally in order to improve the way of release memory, at the moment of releasing we removed all the views inside the scrollviewcontroller. Excuse me for the bad english. If you want I can share with you the class modified, but tomorrow. Here it's 22:52 . Good bye.

**Comment by nin9creative** [comment]
That'd be great!  Would love to see what you had to do...

**Comment by josearellano** [comment]
How can I share what I did?

**Comment by nin9creative** [comment]
Put the file in Dropbox?

Warm Regards,
Andrew

**Comment by josearellano** [comment]
I would like to upload these files to the github, if it's possible.

**Comment by cytec** [comment]
try https://gist.github.com/ or if there are more files i guess you have to create a repo for that...

**Comment by nin9creative** [comment]
@peperepito Can you share how you are instantiating and loading your "BakerViewController" with issues that you have downloaded from your server?  I got the files that you sent that allow Baker to be displayed in a viewcontroller - but am curious on how you are launching the issues that you have downloaded?

**Comment by folletto** [comment]
Hi @nin9creative! 
How is the shelf fork going?

**Comment by nin9creative** [comment]
It's going :) Where I don't know... LOL

We have a pretty functional shelf view with downloads etc working. Mpena has a bunch of InApp purchase code
in a pull request that I need to review and look at. I think there are still some pieces that we need to work out (background downloading, 
push notifications).

We have someone else actually working on a server component to do receipt validation and some other stuff.

Just need to get the time to try to pull it together and see if there are issues.

I think someone else is actually working on a free Non Newstand app that is using the shelf to have free issues/content.

**Comment by folletto** [comment]
Wow. Thanks for the update. :)

I'm asking because if we don't find any issue tomorrow we should be able to release version 3.2 and as such we'll move to version 4.0... shelf by next week. :)

**Comment by mpena2099** [comment]
Nice, I think we need help to finish the shelf. Any help. :)

**Comment by Bobramyl** [comment]
Hello everyone,

I would like to share our progress with shelf / advanced version of reader. I think screens will speak for themselves soo, here you go: http://imgur.com/a/XPBeX#1

I made heavy modifications to Baker base code (which was mostly inevitable to get things rollin'). I also created shelf using AQGridView (well, I had to modify that one too..). As of today, application can:
- Load publication database from remote server (uses core data to store them which allows offline usage)
- Download publications. I wrote entire network on both sides (PHP on server side / Obj-C / ASIHTTP in application), so am not using baker native downloader (I needed a lot more flexibility to work with core data).
- Open any publication. Also book integrity validation is present, if book is corrupted, it will automatically download it again. And it also supports autoupdates from remote server (for now, it update whole book, I might reconsider to differential updating if I had time, which I currently don't).
- New toolbar is added to top of the view. It's customizable and I'm working on functions which will allow to search through entire book, create notes, even draw to views.

Am doing this project for one of the largest university in country, so at the end, there will be all edu materials available (for free). That's why it needs to be that powerful.

As I promised, I will share whatever code I can, but at first, I need to finish it. My deadline is within few weeks.

Hope you enjoyed a little preview.

Sincerely, George

**Comment by mpena2099** [comment]
Nice work @Bobramyl! 

Just one question: you will put any kind of In App Purchase or Newsstand integration?

**Comment by nin9creative** [comment]
Looks pretty good!

Warm Regards,
Andrew

**Comment by mutador** [comment]
but i see you use a very old baker in this case laker framework, it will be a problem integrate a old version of baker o this is only for hpub and the books can use the last version?

always the books must be the last version and the visor a independent code

**Comment by Bobramyl** [comment]
I'm using last baker out there (3.2) and books are not using laker that much, they take some of its parts but they are all made from scratch. Don't worry, I know how to keep code up-to-date :)

As of InApp purchase & newsstand, that's actually pretty easy because I already made those for another App and due to heavy usage of core data, implementation should be walk in the park. BUT the client does not want such integration in this app, so I might provide my inapp class so you can integrate it for yourself (it's really rather easy).

**Comment by nin9creative** [comment]
I think the InApp class would be great to be able to see! :)

**Comment by nurtext** [comment]
@Bobramyl Great work! Can't wait to see the source pulled into the repo. Newsstand would be great and InApp purchases, too… :)

**Comment by Bobramyl** [comment]
I've uploaded inApp purchase manager to GIT, it's singleton and it's pretty easy to use + customisation is only about changing the kickOff method & product identifiers. Also usable in simulator, which is great.

https://gist.github.com/2304762

Hope it will help somehow / somebody :)

**Comment by nurtext** [comment]
@Bobramyl Thanks for sharing! When can we expect the code for the shelf? Can't wait to test it :)

**Comment by Bobramyl** [comment]
I will have to make some changes first due to licensing (I just can't share full code even if I want to - it's huge private project after all), but I will try hard to push anything ASAP. I will most definitely keep you guys informed :)

**Comment by Bobramyl** [comment]
I got yet another update & preview, I've successfuly implemented fullscale search of entire book so that might be very interesting update for baker.

http://imgur.com/duRSf

I will post code for this part within few days.

**Comment by nurtext** [comment]
@Bobramyl Very impressive, keep up the good work! :)

**Comment by folletto** [comment]
That's really impressive. :)
We can't wait to see your contribution. :D

**Comment by mikadonguyen** [comment]
@Bobramyl luv you ;)

**Comment by Bobramyl** [comment]
Hi guys, I've added ability to mark text within any book page you want and store it, so you can create your own marks just like iBooks. The difference is that you can then click send -> highlights and that will take selected texts and fill mail contents with them (might actually have very good use if you want to send citations etc.). Also you can mail current selection (second option), so no highlight is required in this case. Feedback appreciated, also I will wait for your opinions.

http://imgur.com/EYgKE

Regards, JT

**Comment by nurtext** [comment]
@Bobramyl Impressive! You should consider an option, added to book.json, to control if citation/copying is allowed. iBooks disallows copying text regarding copyright issues.

**Comment by Bobramyl** [comment]
That's probably very good idea - i did not consider it though because the application I'm making should allow citations / copying in the first place. But hey, I could add an option, it's like 5 minutes.

**Comment by nurtext** [comment]
+1

**Comment by hebertialmeida** [comment]
+1

## 

Sent from my iPhone

Em 17/04/2012, às 12:49, Cedric Kastner
reply@reply.github.com
escreveu:

**Comment by bahmet** [comment]
+1

**Comment by cptobj** [comment]
+1

**Comment by zonkzen** [comment]
Does anyone know if https://github.com/nin9creative/baker is free for commercial use?

**Comment by folletto** [comment]
As far as I know it's not complete - but I would be very happy to be proven wrong. 
You should probably try to contact him directly. :)

**Comment by zonkzen** [comment]
It's not fully complete, but really ready to be used!
I'll contact him, thanks.

**Comment by nin9creative** [comment]
We have a pretty good branch called bakershelf-newsstandkit that has a lot of the Newsstand features implemented.

https://github.com/nin9creative/baker/tree/bakershelf-newsstandkit

Here is some notes on usage for this branch.

https://github.com/nin9creative/baker/wiki/bakershelf-newsstandkit-Branch-Instructions

Getting it all setup right with the App Store is the biggest pain actually...

Andrew

**Comment by nin9creative** [comment]
Yes, it's free for commercial use...

The non newstand version works pretty well.  The standard Bakershelf has already been used in one
publication that made it to the app store...

Andrew

**Comment by medaura** [comment]
I'm assistant publisher at a small but by no means indie magazine. Our publication spent a lot of money to develop a native iPad app the dinosaur way and we just found out that the development firm that produces it for us will no longer support it in the near future. I've been quietly following Baker's development and if this shelf issue is solved, I'll try to convince my team to reinvent our app in the Baker framework. 

I'm going to look into Andrew's work soon to see if it's suitable for our needs, but I was really wowed by George's (Bobramyl) screenshots. 

George, may I ask how close you are to sharing usable code?

And Davide,

Could you please clarify something for me, in regards to what was already asked by kremisi about bypassing Apple? We would like to transition from in-app purchases to a subscription bundle, whereby all our subscribers get access to our app. Your reply to kremisi was a bit too technical for me to fully wrap my head around. Would you mind spelling it out for me? How could I send a Baker-shelf app to subscribers via email or login, without publishing in the Newsstand or App Store at all? 

If this could be done, I'm confident I can persuade my team to move in this direction, which would make me very happy, as I'm itching to do things this way. Our publisher is considering paying for WoodWing and Adobe DPS, but I think it would be a waste.

Many thanks in advance!

**Comment by Bobramyl** [comment]
It's fully functional with no known bugs and all features implemented, its going to AppStore as we speak, I will then post link for itunes download & some parts of code. Please bear in mind that even I would like to share entire codebase of final version, I'm tightly locked by contract. But I will do my best :)

**Comment by nurtext** [comment]
The contract is useless, what matters is the opensource license under which Baker is released and as far as I know this clearly describes how to handle opensource… ;-)

**Comment by Bobramyl** [comment]
That indeed is, but it all depends on where you live etc. But as I already said, I'd like to share it, so you know :)

**Comment by nurtext** [comment]
Would be great :)

**Comment by folletto** [comment]
@medaura You **can't** bypass Apple. The discussion above was about how to trigger downloads from the web to be handled by a Baker application (and that's done by creating a "dedicated" HTTP protocol). For either subscription or in-app purchases you can do just two things:
1. Use the Apple-provided mechanisms (in-app or subscription).
2. Create a free app that is still useful, and _without any link_ in the app, provide a way to buy from your website.

The second choice is very risky, consider that's part of a long-going fight between Apple and some publishers and it will damage your sales.

~

@nurtext @Bobramyl The point about open source is surely clear and it's surely not dependent of where you live, however we aren't using the GPL license, but the BSD one. BSD doesn't force any release of the code in the open, it's up to the goodwill and choice of the developers that take Baker's open code and build upon it.
We always advise to release in the open, because — big or small companies — there's no harm related to it. :)

**Comment by medaura** [comment]
So Davide,

Are you saying it's impossible to implement a web-app with Baker? Offline storage? It was in that context that I was thinking of web downloads. 

The reason I want to bypass Apple is not about saving money, but giving print subscribers access. Big publishers have apps that allow both subscriptions as in-app purchases, AND validation of their existing subscribers. This second element is a big pain to get together, as it involves API calls to whomever manages your digital entitlements and whatnot. So if you go through the App Store and implement in-app purchases (which we currently are), it's very hard to give existing subscribers access. 

I was thinking that with a web-app behind a login wall on our website, we could get around these headaches. We'd sacrifice revenue from in-app purchases, but that would be negligible. The most important thing for us is to be able to give existing subscribers an iPad version of the magazine.

I've been getting so impatient that I even considered laying out the magazine in iBooks Author, and give subscribers access to the .iba file, but the design limitations of iBooks Author templates are too much to swallow. 

I thought that if the Baker Shelf were pure HTML5, then it could work as a web app and do the trick, but perhaps I am missing something. In any case, if anyone here thinks he could make it work with some custom development, please let me know. We're willing to pay market rate to get it done. And, for a discount, we'd be open to making the code open source. 

Any other tips or advice would be most welcome.

**Comment by folletto** [comment]
I'm saying it's _technically_ possible with Baker, but I think Apple will block you out. They did that with big companies, I don't think they will allow you. :)

**Comment by medaura** [comment]
p.s. a lose example is Aside Magazine, from the maker of Laker. 

http://asidemag.com/

It works as a web-app on the iPad, but it has no shelf. We'd want to do the same but with a shelf for multiple issues.

**Comment by folletto** [comment]
Unfortunately that's not an example (and as far as I know it's not from Laker's author). They built _everything_ in HTML5, they don't publish on the store. So, they don't use anything like Baker (and the speed is very affected).

**Comment by medaura** [comment]
I think I'm beginning to understand now: When I thought of Baker, I focused on the book folder and hpub. But I guess the whole point of Baker is to wrap that into a native app, to go into the app store. Hmmm.... Of course! 

Could what I am talking about be done anyway? A shelf environment with offline storage fully made in HTML5? Could anyone give some tips and direction?

That example <em>is</em> from Laker's author, but he collaborated with someone else.

**Comment by folletto** [comment]
If it's fully HTML5, you can do whatever you want, yes. :)

**Comment by ffranke** [comment]
Hey there,
I'm not the maker of Aside. Johannes Ippen is. I worked with him together on another project (a book), but I was NOT in any way involved in aside.

cheers
florian (Laker)

**Comment by medaura** [comment]
Ah, yes, it was the <em>book</em> that you collaborated on. My bad.

So now I am wondering to what extent I can possibly use Baker for my publication's needs. Is the shelf solution Andrew has built all in HTML5? If so, I could perhaps only use that, and ignore the part of Baker that wraps the shelf into an app.

**Comment by nin9creative** [comment]
What we have done is not built in HTML5.

It's an objective-c wrapper for Baker.  It deals with remotely loading available issues, downloading issues, managing issues on the iPad, Newsstand Subscriptions, In App purchase etc...

We've rewritten Baker just a tiny bit in order to be able to use it as it's own ViewController within the shelf.

There are still some loose ends we are working on (never enough time) but it's close to being somewhat feature complete.

Andrew

**Comment by metalgab88** [comment]
I'm very interested on Bobramyl project cause i've to do exactly the same thing.. it would be great if he try to publish it under open source license and i hope i can develop and extend his project.. i'll stay tuned! 

Great Job everyone!

**Comment by Bobramyl** [comment]
feel free to contact me (info [at] chuckdesign [dot] cz)

**Comment by danupoyner** [comment]
Hi everyone. I have been following this issue with interest. Just wondering if anyone has a realistic estimate of how long until this feature is implemented. I have a digital magazine made with Baker and am currently working on the second edition. Would love to be able to take advantage of the fully-implemented shelf in time for that, but of course this may not be possible. If not, what do people suggest is the best way to release a new edition (ie new Hpub) so that either/both are available to read within a single app?

**Comment by folletto** [comment]
We are almost done with the code refactoring and modularization, leading us next with the Shelf. :)

As you can see from Issue #59 we have merged back to `MASTER` the first iteration of the modular version of Baker (now `BakerViewController` and not anymore `RootViewController`), allowing a greater extent of reusability and implementation in new projects.

Please see Issue #59 and help us to check for bugs so we can safely then work on the Shelf/Multi-issue itself for our next milestone, version 4.0. :)

**Comment by Bobramyl** [comment]
Well played! :)

**Comment by nin9creative** [comment]
I didn't look at the check in yet, did you guys get rid of the separate touch overlay stuff?

**Comment by folletto** [comment]
We tried, but it seems there's not way to remove it while keeping the same interactions due to the fact that `UIWebView` "eats" the events.
It's still there, we managed to make it lighter and self-contained tho. :)

The idea is that now you can just initialize it as a class, and it works. :)

**Comment by Bobramyl** [comment]
Naw, it's really not possible, becouse there is no way to subclass UIWebView so you can add your own touch handlers / gesture recognizers (a lot of things will sieze to work). I had to make workaround with another UIView to get user interaction above the UIWebView and it's still not 100%valid solution, but hey, works.

As of update, I've checked it and it looks sweet, keep up good job guys :)

**Comment by folletto** [comment]
Given the results from the refactoring, we now started working on the shelf. We studied all the different approaches that have been done or tried up until now and while each of them is too far away from the core of functionalities of Baker to be merged easily on our side, we learned quite a lot from each approach, the various problems (appearing in the threads, commits and emails) and that allowed us to have a speed start on it. :)

So, **thank you all for your efforts!**

Where are we? Well, this is the very beginning and you can follow the branch `shelf` to be updated on that. It's still **NOT WORKING**, so be aware that until it gets merged in `MASTER` it can explode, turn off the lights, burn your bacon and steal your cookies:
https://github.com/Simbul/baker/tree/shelf

Use this _only_ to track the progress and - if you want - coordinate with us to provide patches and updates. Please avoid also to filing bugs until we can actively look for them. ;)

We'll update this thread. :)

**Comment by mpena2099** [comment]
Hi @folletto!

Why not just use the code from https://github.com/nin9creative/baker ?

**Comment by folletto** [comment]
We studied the code quite a bit, actually it has been very useful and helped us a lot.

We were very sad but we had to decide against starting from that one because:
1. The last merge of the features was done on the 10th of December given the git log, so there are loads of commits, features and fixes missing.
2. You added a lots of features and we aren't quite able to determine which one is where and which can be merged back and which is instead customized for your personal needs.
3. We never got the message that it was complete, nor a push request of features or patches... but it might be that we missed that. We can't merge back features that aren't "complete" because then we will have to figure out why ourselves from code that for us is unknown.

So... we'll be happy to accept contributions if you are able to split them out in specific features updated on the current codebase... otherwise we'll keep checking the code to see your decision and choices and see what we can learn from your progress. :)

**Comment by nin9creative** [comment]
We've been working a bunch off the bakershelf-newsstandkit branch and much of that hasn't been merged into the bakershelf branch itself.  Mostly because this newsstand stuff is complicated.  I wish I had more time to work on it but have been knee deep in multiple projects lately!  However from a shelf point of view minus the newsstand integration, much of it works pretty good.

**Comment by folletto** [comment]
So... for the weekly progress update: there are still a few details to sort out, but the current `shelf` branch now supports multiple books present already in-app (if a folder is present in the books folder, it will be read as multiple issues).

It's still completely local (no support for download and probably some related bug), but the read-open-close part should work.

**Comment by drking0211** [comment]
Sounds as though you are getting there with the shelf support and the multi issue problem.  

I have spent the last month building a front end for the Baker framework whereby our Editors access a web portal and submit content (text, pictures, videos etc) to the back end SQL system which then auto generates the HTML pages; auto generates the TOC and Index for navigation and then moves all the various folders and files into the correct syntax.  

Our Editors are over the moon with the new system as they are taken entirely out the tech loop and are able to produce apps in just a few hours which are error free ( well, for the most part), look beautiful and are consistent.

But Apple's remit for the Newsstand is fairly strict so I am anxious to try and get this worked out.

So any news on the 4.0 release would be very welcome.

Great job on the Framework.  Best out there in my opinion, by miles.

**Comment by folletto** [comment]
Thanks! :)

A little update: now the current `shelf` branch supports multi-book _locally_ and the shelf + toolbar is skinnable (three images are there for this reason to be customized). This part is already in place and working: just drop multiple books in the `books/` folder and Baker picks them up.

It's still missing the download feature (and as such, the `book://` protocol is currently broken) and a list of other smaller issues here and there to be fixed (no special support when just one issue is present, some visual issues when transitioning, titles to be added, ...).

**Comment by mpena2099** [comment]
Hi folks!

Please take a look on the new version of my "BakerShelf" APP in http://itunes.apple.com/us/app/revista-canamix/id498066240?mt=8, using https://github.com/nin9creative/baker.

It's working, but we still have some bugs to fix.

@folletto, Baker group will work in its own solution or wanna be cool I try to merge the code from https://github.com/nin9creative/baker here???

Thanks!

**Comment by folletto** [comment]
@mpena2099 for us it would be great to incorporate your shelf even if now we started developing our own. The most important thing however is that you should first integrate the current master changes in your branch to avoid repeating bugs that were already fixed (your latest update from us was on the 10th of December last year), make sure that _everything_ works on Baker as before (a few weeks ago we tested it and we noticed some differences in the touch events on the books) and then we can see what is the best way to merge.

We can't accept it as a whole if it's not on feature parity with the core Baker, you can understand. :)

On the other side, we might also bring in just some specific parts: for example, right now we have the shelf up but not the remote download and we will have to implement two different classes probably for users that want to use Newsstand and users that want to have the app as a standard app. It would be _AMAZING_ if you could extract the code that handles Newsstand downloads so we can plug it in. :)

**Comment by nin9creative** [comment]
Nice work @mpena2099!!!

**Comment by reecegeorge** [comment]
hi guys, i am new to github and wanted to ask a basic question about using mpena2099 newsstand solution.

is mpena2099 newsstand solution the same as https://github.com/nin9creative/baker ?

is mpena2099's code available somewhere?

thanks

**Comment by nin9creative** [comment]
He forked mine and has been pushing some fixes and changes back to it.

**Comment by mpena2099** [comment]
@folletto, I'm a little busy right now, but I'll try to put the shef from nin9creative/baker to Baker itself.

@nin9creative, do you agree with this? Because... it will be the beginning of killing nin9creative/baker shelf. :)

**Comment by nin9creative** [comment]
Yeah, we do need to update it to use the core baker updates...  We are definitely lagging behind a bit on that as things had to change a bit to get it to work....

**Comment by reecegeorge** [comment]
if mpena299 has the code working, and is too busy to work on it, why not just upload it, so we can work on it?

am i missing something?

**Comment by folletto** [comment]
...just that the code is already a fork on GitHub, see the link above. ;)

If you see the network (link in the menu at the top), you'll notice - other than Baker and a few others - the excellent collaboration on @nin9creative's tree and @mpena2099 pushing to it. :)

**Comment by mpena2099** [comment]
Yes, thats it. See https://github.com/nin9creative/baker/network @reecegeorge.

**Comment by reecegeorge** [comment]
this is very cool

**Comment by folletto** [comment]
Tonight we broke down the remaining steps for us to proceed with the shelf.
We split in two: first getting the shelf working with local issues, then getting it working with remote issues (NewsstandKit).

CHECKLIST CONSOLIDATE:
~~1. Add: The book should derive the folders from the path it was initialized from~~
~~2. Add: Write the book status when the book is closed~~
~~3. Fix: Animation glitch when opening a book from the shelf: show the bar on top open, then close it~~
~~4. Fix: Check the screenshot glitch~~
~~5. Add: Book "cover" viewed as thumbnail on the shelf~~
~~6. Add: switch on SHELF if there's only one book, open directly (take out ShelfManager)~~
~~7. Chore: Cleanup (downloader, sha, …)~~

CHECKLIST SHELF:
A. Add: preprocessor code to switch NK on
B. Add [NK]: download list of issues from server, visualize list on shelf, save locally
C. Add [NK]: download button, download issue from shelf
D. Add [NK]: delete issue

**Comment by Simbul** [comment]
> 5 . Add: Book "cover" viewed as thumbnail on the shelf

Implemented in a8ae6472bbfa7a2b50467edcdf3eccde4db31b9b

**Comment by Simbul** [comment]
> 1 . Add: The book should derive the folders from the path it was initialized from

Started implementing in c113173cdec81f00fae1d511b598a248f032b133, with an `ID` for `BakerBook`. The `ID` can then be used as a base path for all the Baker "internal" folders.

**Comment by Simbul** [comment]
Got to a first working implementation in a8e2d0658d4e836cf44bf24529d2d058e0db4c4b, at least for the screenshots path.

It's worth noting that the screenshots folder is not based on the path the book was initialised from, but on a folder based on the book `ID`.

**Comment by Simbul** [comment]
> 1 . Add: The book should derive the folders from the path it was initialized from

Basic implementation completed in 053c592, for screenshots and book status.

**Comment by Simbul** [comment]
> 2 . Add: Write the book status when the book is closed

Completed in 53c3a132066d5288aee63e52a98f7216724d7ebf.

> 3 . Fix: Animation glitch when opening a book from the shelf: show the bar on top open, then close it
> 4 . Fix: Check the screenshot glitch
> 6 . Add: switch on SHELF if there's only one book, open directly (take out ShelfManager)
> 7 . Chore: Cleanup (downloader, sha, …)

Done in various commits, the last of which is 16df9640a26c6b3949f02799e830cc25cb2a0f0b.

**Comment by reecegeorge** [comment]
@Simbul good work

**Comment by danupoyner** [comment]
I've been testing the Shelf code (ae6b631) and it seems to be coming together well. However, when I use my own hpubs it crashes when it loads... I'm still quite n00bish with all this but it seems like it's choking on something in the .json file. Have the requirements changed?

Also, have you settled on image dimensions for the covers? I notice the latest commits have changed them from 100x150.

Cheers,
Danu

2012-10-19 17:12:14.625 Baker[369:707] JSON Parsed successfully, book "the-canny-outlaw-issue-1-7869d1dffce2c1f3 - The Canny Outlaw - Issue 1" created
2012-10-19 17:12:14.627 Baker[369:707] • INIT
2012-10-19 17:12:14.630 Baker[369:707]     Device Width: 768.000000
2012-10-19 17:12:14.631 Baker[369:707]     Device Height: 1024.000000
2012-10-19 17:12:14.651 Baker[369:707] Error when loading book status: Error Domain=NSCocoaErrorDomain Code=260 "The operation couldn’t be completed. (Cocoa error 260.)" UserInfo=0xc657f70 {NSFilePath=/var/mobile/Applications/A54B758A-DD2C-4D72-86CC-4E26615607C0/Library/Caches/statuses/the-canny-outlaw-issue-1-7869d1dffce2c1f3.json, NSUnderlyingError=0xc657ec0 "The operation couldn’t be completed. No such file or directory"}
2012-10-19 17:12:14.655 Baker[369:707] STATUS: page: (null)
2012-10-19 17:12:14.658 Baker[369:707] STATUS: scrollIndex: (null)
2012-10-19 17:12:14.661 Baker[369:707] • LOAD BOOK WITH PATH: /var/mobile/Applications/A54B758A-DD2C-4D72-86CC-4E26615607C0/Library/Caches/Newsstand/2A7AB134-CD5F-4686-9CF6-7B5C44F8A68B
2012-10-19 17:12:14.664 Baker[369:707] • Reset leftover page slot
2012-10-19 17:12:14.667 Baker[369:707] • Reset page details array and empty screenshot directory
2012-10-19 17:12:14.673 Baker[369:707]     Pages in this book: 16
2012-10-19 17:12:14.674 Baker[369:707] Screenshots are stored in /var/mobile/Applications/A54B758A-DD2C-4D72-86CC-4E26615607C0/Library/Caches/screenshots/the-canny-outlaw-issue-1-7869d1dffce2c1f3
2012-10-19 17:12:14.685 Baker[369:707] • Set size for orientation: portrait
2012-10-19 17:12:14.686 Baker[369:707] • Set tappable area size
2012-10-19 17:12:14.688 Baker[369:707] • Init page details for the book pages
2012-10-19 17:12:14.820 Baker[369:707]     Prevent page from changing until layout is updated
2012-10-19 17:12:14.822 Baker[369:707] • Show page details for the book pages
2012-10-19 17:12:14.826 Baker[369:707]     Unlock page changing
2012-10-19 17:12:14.827 Baker[369:707] Set IndexView size to 768x1024, with pageY set to -20
2012-10-19 17:12:15.133 Baker[369:707] Path to index view is /var/mobile/Applications/A54B758A-DD2C-4D72-86CC-4E26615607C0/Library/Caches/Newsstand/2A7AB134-CD5F-4686-9CF6-7B5C44F8A68B/index.html
2012-10-19 17:12:15.137 Baker[369:707] Path to index view is /var/mobile/Applications/A54B758A-DD2C-4D72-86CC-4E26615607C0/Library/Caches/Newsstand/2A7AB134-CD5F-4686-9CF6-7B5C44F8A68B/index.html
2012-10-19 17:12:15.140 Baker[369:707] • Add page to the loding queue
2012-10-19 17:12:15.142 Baker[369:707] • Add page to the loding queue
2012-10-19 17:12:15.144 Baker[369:707] • Handle loading of slot 0 with page 1
2012-10-19 17:12:15.145 Baker[369:707] • Setup new page for loading
2012-10-19 17:12:15.155 Baker[369:707] • Setup webView
2012-10-19 17:12:15.204 Baker[369:707] • Loading: book/01_cover.html
2012-10-19 17:12:15.898 Baker[369:707] • Should webView load the page ?
2012-10-19 17:12:15.900 Baker[369:707]     Page is current page and current page IS delaying loading --> load page
2012-10-19 17:12:15.905 Baker[369:707] • Page did start load
2012-10-19 17:12:16.399 Baker[369:707] • Page did finish load
2012-10-19 17:12:16.410 Baker[369:707] • Setting current page height from 0 to 1024.000000
2012-10-19 17:12:16.414 Baker[369:707] • webView hidden: 0 animating: 1
2012-10-19 17:12:16.417 Baker[369:707] • Handle loading of slot 1 with page 2
2012-10-19 17:12:16.419 Baker[369:707] • Setup new page for loading
2012-10-19 17:12:16.427 Baker[369:707] • Setup webView
2012-10-19 17:12:16.431 Baker[369:707] • Loading: book/02_contents.html
2012-10-19 17:12:16.832 Baker[369:707] -[NSNull integerValue]: unrecognized selector sent to instance 0x3ee645e0
2012-10-19 17:12:16.834 Baker[369:707] **\* Terminating app due to uncaught exception 'NSInvalidArgumentException', reason: '-[NSNull integerValue]: unrecognized selector sent to instance 0x3ee645e0'
**\* First throw call stack:
(0x33baf88f 0x326a1259 0x33bb2a9b 0x33bb1915 0x33b0c650 0x84431 0x35974c01 0x33bb27e4 0x33b0d7b1 0x33b0d3cf 0x305d8e9f 0x33bb1a83 0x33b0c650 0x33bb27e4 0x33b0d7b1 0x37332649 0x37345b45 0x33b83ad3 0x33b8329f 0x33b82045 0x33b054a5 0x33b0536d 0x33083439 0x357fae7d 0x772ff 0x772a0)
terminate called throwing an exception(lldb)

**Comment by folletto** [comment]
Uhm no nothing changed on the Hpub standard, so it's likely a bug of some kind.
Thank you very much for the detailed report, it helps a lot. :)

**Comment by folletto** [comment]
Btw, we are almost there with the `shelf` branch. Have a look and test it if you can, so we can find problems like the one above (thanks again). :)

**Comment by danupoyner** [comment]
Another question... the Shelf now displays the app in Newsstand. My understanding is for an app displayed in Newsstand to be approved by Apple it has to have in-app purchase. What's your position with this? Will the guts of the shelf still work without the flag to display in Newsstand?

**Comment by folletto** [comment]
To use Newsstand features you need to be in Newsstand, but in-app purchase is a different thing and will not be included in 4.0 (unless there's a blocker we aren't aware of). If you know differently please link us a resource so we can review it. :)

To explain our thinking better:
1. We are trying to have a 4.0 that will support either **standalone apps without download** (containing multiple books from the start) or **Newsstand apps with download** (with background download, remote server and remote updates).
2. This works for free publications. If you need to sell them you have to implement in-app purchase yourself, it will not be part of Baker 4.0.

**Comment by danupoyner** [comment]
Ok, that helps. To clarify - I can either bundle the multiple issues with the app and release an app update to put out a new issue, or I can use Newsstand to push out updates remotely but I need to build on the 4.0 code myself to add the required features? There is no middle option whereby I push out issue updates remotely but the app is not Newsstand enabled?

**Comment by folletto** [comment]
That's correct. :)

**Comment by folletto** [comment]
Ok, we can close this issue now:
1. Baker in the branch `shelf` **manages multiple books**.
2. Baker in the branch `shelf` has a **shelf** to visualize them.

We are closing this issue also because it's getting too big and the discussion is a bit too wide. This issue was quite narrow in scope originally (referring just multiple books and shelf, bundled inside the app) while the rest was split in other issues.

So, to follow the progress from here, the other two issues are:
1. Issue #100, where we track the implementation of Newsstand.
2. Issue #519, where we track the implementation of the iPhone shelf (right now it's iPad only).

Plus of course the whole Milestone 4.0: https://github.com/Simbul/baker/issues?milestone=8&state=open

This means that the branch `shelf` is still in development, so if you can help testing it, it would be great! :+1: 

—

So, how Baker handles a shelf without a full Newsstand implementation right now? It's easy: as it used to work before, it bundles the book inside the app. It's not a great solution for everyone, we understand, but it's a request we received a few times and so there it is.

We explain the two modes in this wiki page: https://github.com/Simbul/baker/wiki/Newsstand-vs-Bundled-publications-support-in-Baker-4.0

—

Thanks everyone for the support, it has been a great discussion. :)
